be/src/runtime/tablets_channel.cpp

```diff
@@ -49,115 +56,162 @@ TabletsChannel::~TabletsChannel() {
+        response->mutable_status()->add_error_msgs(
+                fmt::format("invalid sender_id {} in PTabletWriterAddChunkRequest, limit={}", request.sender_id(),
+                            _senders.size()));
+        return;
+    }
+    if (UNLIKELY(!request.has_packet_seq())) {
+        response->mutable_status()->set_status_code(TStatusCode::INVALID_ARGUMENT);
+        response->mutable_status()->add_error_msgs("no packet_seq in PTabletWriterAddChunkRequest");
+        return;
     }
 
-    auto& pchunk = params.chunk();
-    {
-        std::lock_guard<std::mutex> l(_global_lock);
-        if (_chunk_meta.types.empty()) {
-            RETURN_IF_ERROR(_build_chunk_meta(pchunk));
-        }
+    Sender& sender = _senders[request.sender_id()];
+    std::lock_guard l(sender.lock);
+
+    if (sender.next_seq < 0) {
+        response->mutable_status()->set_status_code(TStatusCode::INTERNAL_ERROR);
+        response->mutable_status()->add_error_msgs("Tablet channel has been cancelled");
+        return;
+    } else if (request.packet_seq() == sender.next_seq) {
+        sender.next_seq++;
+    } else if (request.packet_seq() < sender.next_seq) {
+        LOG(INFO) << "Ignore outdated request from " << cntl->remote_side() << ". seq=" << request.packet_seq()
+                  << " expect=" << sender.next_seq << " load_id=" << _load_channel->load_id();
+        response->mutable_status()->set_status_code(TStatusCode::OK);
+        return;
+    } else {
+        LOG(WARNING) << "Out-of-order request from " << cntl->remote_side() << ". seq=" << request.packet_seq()
+                     << " expect=" << sender.next_seq << " load_id=" << _load_channel->load_id();
+        response->mutable_status()->set_status_code(TStatusCode::INVALID_ARGUMENT);
+        response->mutable_status()->add_error_msgs("out-of-order request");
+        return;
     }
 
-    vectorized::Chunk chunk;
-    RETURN_IF_ERROR(chunk.deserialize((const uint8_t*)pchunk.data().data(), pchunk.data().size(), _chunk_meta,
-                                      pchunk.serialized_size()));
-    DCHECK_EQ(params.tablet_ids_size(), chunk.num_rows());
-
-    size_t channel_size = _tablet_id_to_sorted_indexes.size();
-    std::vector<uint32_t> row_indexes(chunk.num_rows());
-    std::vector<uint32_t> channel_row_idx_start_points(channel_size + 1);
-    {
-        // compute row indexes for each channel
-        channel_row_idx_start_points.assign(channel_size + 1, 0);
-        for (uint32_t i = 0; i < params.tablet_ids_size(); ++i) {
-            uint32_t channel_index = _tablet_id_to_sorted_indexes[params.tablet_ids(i)];
-            channel_row_idx_start_points[channel_index]++;
-        }
+    auto res = _create_write_context(request, response, done);
+    if (!res.ok()) {
+        res.status().to_protobuf(response->mutable_status());
+        return;
+    } else {
+        response->mutable_status()->set_status_code(TStatusCode::OK);
```

Comment:
Better to set ok when add_chunk is finished

Reply:
It's an optimization to reduce lock contention

Reply:
I've added some comments to explain this, PTAL @wyb